### PR TITLE
[metrics] Add breakdowns for when tasks may be blocked during scheduling and in user code

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This PR:
- Refactors and expands the task states enum (see common.proto for a holistic summary of the proposed task states)
- Adds additional breakdowns to track Ray task state when blocked on fetching dependencies in the object manager, and blocked on user calls to ray.get/wait during task runtime.

### CODE REVIEW

File: src/ray/core_worker/core_worker.h [hunk 1]
```diff
@@ -59,6 +59,97 @@ namespace core {
 
 JobID GetProcessJobID(const CoreWorkerOptions &options);
 
+/// Simple container for per function task counters. The counters will be
+/// keyed by the function name in task spec.
+struct TaskCounter {
```
Comment (ericl) [inline src/ray/core_worker/core_worker.h:64]: Moved from below for visibility in core_woker.cc

**Comment by ericl** [comment]
Ping @rickyyx @rkooo567 for review on this.

File: src/ray/core_worker/core_worker.h [hunk 2]
```diff
@@ -59,6 +59,99 @@ namespace core {
...
+  mutable absl::Mutex tasks_counter_mutex_;
+  absl::flat_hash_map<std::string, int> pending_tasks_counter_map_
+      GUARDED_BY(tasks_counter_mutex_);
+  absl::flat_hash_map<std::string, int> running_tasks_counter_map_
+      GUARDED_BY(tasks_counter_mutex_);
+  absl::flat_hash_map<std::string, int> finished_tasks_counter_map_
+      GUARDED_BY(tasks_counter_mutex_);
+
+  /// Tracks the total number of running tasks on this process.
+  int64_t running_total_ GUARDED_BY(tasks_counter_mutex_) = 0;
```
Comment (ericl) [inline src/ray/core_worker/core_worker.h:80]: This is getting a bit complex; but in a follow-up PR, I have a cleanup that refactors out a common Counter to simplify across the different locations we're counting tasks by key.

File: src/ray/core_worker/core_worker.h [hunk 3]
```diff
@@ -59,6 +59,99 @@ namespace core {
...
+    UpdateStats();
+  }
+
+  void UpdateStats() EXCLUSIVE_LOCKS_REQUIRED(&tasks_counter_mutex_) {
+    // Note that we set a Source=executor label so that metrics reported here don't
+    // conflict with metrics reported from task_manager.cc.
+    ray::stats::STATS_tasks.Record(
+        running_total_ - running_in_get_total_ - running_in_wait_total_,
+        {{"State", rpc::TaskStatus_Name(rpc::TaskStatus::RUNNING)},
+         {"Source", "executor"}});
```
Comment (rickyyx) [inline src/ray/core_worker/core_worker.h:139]: nit: I wonder if we should have some place/process to document all the labels on the metric? E.g. when it was declared.

File: src/ray/core_worker/core_worker.cc [hunk 4]
```diff
@@ -55,6 +55,30 @@ JobID GetProcessJobID(const CoreWorkerOptions &options) {
 
 namespace {
 
+// Implements setting the transient RUNNING_IN_RAY_GET and RUNNING_IN_RAY_WAIT states.
+// These states override the RUNNING state of a task.
+class ScopedTaskMetricSetter {
+ public:
+  ScopedTaskMetricSetter(WorkerContext &ctx, TaskCounter &ctr, rpc::TaskStatus status)
```
Comment (rickyyx) [inline src/ray/core_worker/core_worker.cc:62]:
```suggestion
  ScopedTaskMetricSetter(const WorkerContext &ctx, TaskCounter &ctr, rpc::TaskStatus status)
```

**Review by rickyyx — APPROVED** [review]
LGTM - left a question on docs. 
Appreciate the added clarity on task scheduling states transition.

File: src/ray/core_worker/core_worker.h (hunk 3 above)
Comment (ericl) [reply src/ray/core_worker/core_worker.h:139]: Added some comments in metric_defs.cc documenting the STATS_task definition and its labels.

**Review by scv119 — APPROVED** [review]

File: src/ray/object_manager/pull_manager.h [hunk 5]
```diff
@@ -265,6 +265,21 @@ class PullManager {
     std::set<uint64_t> active_requests;
     std::set<uint64_t> inactive_requests;
 
+    void RecordTaskMetrics() const {
+      ray::stats::STATS_tasks.Record(
+          -static_cast<int64_t>(requests.size()),
```
Comment (rkooo567) [inline src/ray/object_manager/pull_manager.h:270]: If the task is spilled to another node before pulling is completed (this is possible when `SpillWaitingTasks` is called IIRC), isn't this double-subtracting (or subtracting multiple times whenever spill happens) the `PENDING_NODE_ASSIGNMENT`state? Maybe we should add up these states again when they are spilled to other node via `SpillWaitingTasks`?

Comment (rkooo567) [reply src/ray/object_manager/pull_manager.h:270]: The repro is not very trivial. Maybe we need to
1. add delay to concurrent object pulling (which can happen upon memory pressure, but we can use the chaos testing tool here)
2. local node has enough CPUs and not enough object store memory for additional pulling (so pulling is inactive)
3. and a remote node has enough object store memory

node 1
    Object store memory: 100MB
    CPUs: 16
node 2
    Object store memory: 100MB

request A: 1 CPU, 90MB object store memory (goes to node 1)
request B: 1 CPU, 90MB object store memory (goes to node 1)

and make request A take a long time, then this is triggered IIRC

Comment (ericl) [reply src/ray/object_manager/pull_manager.h:270]: It doesn't really matter, since the metrics are eventually consistent...

**Pull request merged**